// @cloudbase/framework-core/lib/plugin-sevice-api/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const link_1 = require("../utils/link");
/**
 * 插件服务注入为插件提供的 API
 */
class PluginServiceApi {
    constructor(pluginManager) {
        this.pluginManager = pluginManager;
    }
    /**
     * 获取 manager 实例
     */
    get cloudbaseManager() {
        return this.pluginManager.context.cloudbaseManager;
    }
    /**
     * 获取项目根路径
     */
    get projectPath() {
        return this.pluginManager.context.projectPath;
    }
    /**
     * 项目环境id
     */
    get envId() {
        return this.pluginManager.context.envId;
    }
    /**
     * 获取 logger 实例
     */
    get logger() {
        return this.pluginManager.context.logger;
    }
    /**
     * 获取资源操作 API
     */
    get resourceProviders() {
        return this.pluginManager.context.resourceProviders;
    }
    /**
     * 获取云开发 CLI 配置文件
     */
    get projectConfig() {
        return this.pluginManager.context.projectConfig;
    }
    /**
     * 生成命令行链接
     */
    get genClickableLink() {
        return link_1.genClickableLink;
    }
}
exports.default = PluginServiceApi;
